load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java

```diff
@@ -0,0 +1,108 @@
+
+    private final LinkedHashMap<Long, Long> failedInstances = new LinkedHashMap<>();
+    private volatile ServiceInstanceWithStatGathering lastSelected;
+
+    public StickyLoadBalancer(Duration failureBackOff) {
+        this.failureBackOffNs = failureBackOff.toNanos();
+    }
+
+    @Override
+    public ServiceInstance selectServiceInstance(Collection<ServiceInstance> serviceInstances) {
+        if (serviceInstances.isEmpty()) {
+            throw new NoServiceInstanceFoundException("No service instance found");
+        }
+
+        if (lastSelected != null) {
+            for (var instance : serviceInstances) {
+                if (instance.getId() == lastSelected.getId()) {
+                    return lastSelected;
+                }
+            }
+        }
+        Map<Long, ServiceInstance> instanceMap = serviceInstances.stream()
+                .collect(Collectors.toMap(ServiceInstance::getId, i -> i));
+        lastSelected = selectNextInstance(instanceMap);
+        return lastSelected;
+    }
+
+    private ServiceInstanceWithStatGathering selectNextInstance(Map<Long, ServiceInstance> serviceInstances) {
+        for (ServiceInstance serviceInstance : serviceInstances.values()) {
+            if (!failedInstances.containsKey(serviceInstance.getId())) {
+                return new ServiceInstanceWithStatGathering(serviceInstance, this);
+            }
+        }
+
+        Iterator<Map.Entry<Long, Long>> failedInstanceIterator = failedInstances.entrySet().iterator();
+
+        long now = System.nanoTime();
+        while (failedInstanceIterator.hasNext()) {
+            Map.Entry<Long, Long> oldestFailedInstance = failedInstanceIterator.next();
+
+            Long instanceId = oldestFailedInstance.getKey();
+            Long lastFailureTime = oldestFailedInstance.getValue();
+            if (now - lastFailureTime < failureBackOffNs) {
+                break;
+            }
+            failedInstanceIterator.remove();
+            ServiceInstance serviceInstance = serviceInstances.get(instanceId);
+            if (serviceInstance != null) {
+                lastSelected = new ServiceInstanceWithStatGathering(serviceInstance, this);
+                return lastSelected;
+            }
+        }
+        throw new NoAcceptableServiceInstanceFoundException("Each of the available service instances failed " +
+                "within the configured failure-back-off-time");
+    }
+
+    @Override
+    public void recordEnd(long serviceInstanceId, Throwable error) {
+        recordEndAtTime(serviceInstanceId, error, System.nanoTime());
+    }
+
+    // exposed for tests only
+    void recordEndAtTime(long serviceInstanceId, Throwable error, long currentTime) {
+        if (error != null) {
+            failedInstances.put(serviceInstanceId, currentTime);
+            lastSelected = null;
+        }
+    }
+
+    @Deprecated // for tests only
+    LinkedHashMap<Long, Long> getFailedInstances() {
+        return failedInstances;
```

Comment:
to avoid misusage, should we return a copy?

Reply:
done